Fix ClosureComponent to work with SSE alignment requirements

brechtvl opened the pull request:
## Description

In #740 the `alignment` parameter was removed from `register_closure`. This was originally added in bcaf5778, and Cycles still needs 16 bytes alignment for SSE types, while it is fixed to 8 bytes in OSL master now.

This patch uses the simple solution of enforcing 16 byte alignment for `ClosureComponent.mem`, under the assumption that in practice this is the most alignment you need on x86.

Alternatively the `alignment` parameter could be added back. Maybe in theory someone might care about 32 byte alignment for AVX. In practice I guess it's not so likely since alignment is only required for older CPUs that don't support unaligned loads well.

## Checklist:

- [x] I have read the [contribution guidelines](../CONTRIBUTING.md).
- [x] I have previously submitted a [Contributor License Agreement](http://opensource.imageworks.com/cla/).
- [x] I have updated the documentation, if applicable.
- [ ] I have ensured that the change is tested somewhere in the testsuite (adding new test cases if necessary).
- [x] My code follows the prevailing code style of this project.

fpsunflower commented:
This seems like the most expedient fix, but it does mean the minimum size of a closure grows to 32 bytes instead of 24 (with 8 bytes being wasted on alignment in a way no one can use, even if they have weaker alignment requirements).

Looking at this again, I think the more correct way to fix this would be require closure structs to inherit from `OSL::ClosurePrimitive`. This should result in the same memory layout, but the compiler would be able to guarantee alignment itself. Then, because we store its `alignof` as well as `sizeof` when we register a closure -- it would be allocated correctly at runtime.

What do you think?  (It's obviously a much more intrusive change in terms of backwards compatibility)

brechtvl commented:
The size of `ClosureComponent` is reduced to 16 bytes in the updated patch, and then memory after it would be aligned to 16 bytes too.

Inheriting from `OSL::ClosureComponent` is not ideal for SA Arnold, where closures can be allocated by both OSL and C++ shaders. We can probably find some way to do it but it would make things more complicated.

fpsunflower commented on src/include/OSL/oslclosure.h:
```diff
@@ -118,25 +118,28 @@ struct OSLEXECPUBLIC ClosureColor {
+    /// struct. Alignment is forced to 16 bytes to match alignment requirements
+    /// of 64 bit pointers and 128 bit SSE types in user's structs.
+    static const size_t alignment = 16;
+    static size_t aligned_size() {
```
I don't think this is needed, right ? If a type has `alignas(N)` then its `sizeof()` will be at least `N` (so that an array of said struct would have properly aligned entries).

fpsunflower commented on src/include/OSL/oslclosure.h:
```diff
@@ -118,25 +118,28 @@ struct OSLEXECPUBLIC ClosureColor {
+        return (sizeof(ClosureComponent) + alignment - 1) & ~(alignment - 1);
+    }
+
+    void *data () { return (char*)this + aligned_size(); }
```
`(char*) (this + 1)` should work too I think.

fpsunflower commented:
This is even easier. I like the patch (modolo the notes above).

fpsunflower commented:
Longer term, letting user structs inherit from `OSL:ClosurePrimitive` does seem a bit more sound, though having reviewed the code - it does require a bunch of changes. We don't currently store `alignof` the various types like I thought we did.

brechtvl commented:
Simplified the code now as suggested.

I also added a workaround for a GCC bug, to fix the CI failures.
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=69585

fpsunflower commented on src/include/OSL/oslclosure.h:
```diff
@@ -118,25 +118,23 @@ struct OSLEXECPUBLIC ClosureColor {
+    /// Handy method for getting the parameter memory as a void*. Allocation
+    /// allocation will be scaled to requirements of the registered closure
+    /// struct. Alignment is forced to 16 bytes to match alignment requirements
+    /// of 64 bit pointers and 128 bit SSE types in user's structs.
```
64 bit pointers have `alignof == 8`

fpsunflower commented on src/include/OSL/oslclosure.h:
```diff
@@ -118,25 +118,23 @@ struct OSLEXECPUBLIC ClosureColor {
-    ///
-    void *data () { return &mem; }
-    const void *data () const { return &mem; }
+    /// Handy method for getting the parameter memory as a void*. Allocation
```
Typo: "Actual allocation" ?

fpsunflower commented on src/include/OSL/oslconfig.h:
```diff
@@ -131,4 +131,11 @@ using OIIO::string_view;
 #  define OSL_DEPRECATED(msg)
 #endif
 
+// Work around bug in GCC with mixed __attribute__ and alignas parsing
```
Maybe leave a link to the gcc bug you found? (In the distant future maybe we'll be able to re-evaluate if its still needed)

brechtvl commented:
More detailed comments added now.

fpsunflower commented:
LGTM

lgritz commented:
LGTM2. Merging.

brechtvl commented:
Can this be backported to RB-1.9? We need it for Cycles.

lgritz commented:
yes, running tests then will merge it

lgritz commented:
merged into RB-1.9
